### DISCUSSION TITLE
change favicon with mastodon 4

### DISCUSSION THREAD

**Discussion by pinkprius** [mastodon/mastodon#20945] [General]
How do I change the favicon on mastodon 4? Replacing favicon.ico in live/public/, then precompile and restart worked for mastodon 3 but doesn't work since I upgraded.

**Comment by Der-K-2000** [comment]
There are a couple of more favicons in `/home/mastodon/live/public/packs/media/icons`

![grafik](https://user-images.githubusercontent.com/7254827/202571111-5abdac4f-2060-4253-bc6c-a48490268c81.png)

![grafik](https://user-images.githubusercontent.com/7254827/202571286-dff4ecf1-0bb3-4ac7-8f16-c766ad942022.png)

...so only change /favicon.ico wouldn't fix it.

**Reply by pinkprius** [reply]
Thank you so much, it worked!! 

Also to anyone finding this: with Gimp I used Colors -> Colorize to change the colors of the icons!

**Comment by pacohope** [comment]
Here's a more comprehensive solution that worked for me. I put my base file in `/home/mastodon/favicon/favicon.png`. I made the following script as `make-icons.sh`. So my method was:

```
bash make-icons.sh
RAILS_ENV=production bundle exec rake tmp:cache:clear
RAILS_ENV=production bundle exec rails assets:precompile
# then become root
sudo systemctl reload mastodon-web
```

Here's the **make-icons.sh** bash script that converts a single image into all the various sizes/flavours.

```
#!/bin/bash

SOURCE="/home/mastodon/favicon/favicon.png"

cd /home/mastodon/live/app/javascript/icons

for prefix in android-chrome apple-touch-icon favicon
do
  for icon in $prefix-*
  do
    suffix=${icon#$prefix-}
    resolution=${suffix%.png}
    echo -n "."
    convert -scale $resolution ${SOURCE} "${icon}"
  done
done
convert -scale 48x48 ${SOURCE} /home/mastodon/live/public/favicon.ico
echo " done"
```

All those icons are stored in the git repo. So if you change them in `app/javascript/icons` with this script, they will be a change you can either `git stash` or somehow apply programmatically with a patch/branch or other `git` mechanism. Might make it easier to persist than modifying them in `~/live/public`.

Small update because I couldn't figure out what generates `live/public/favicon.ico` so I added a one-off line in the script to deal with that.

**Comment by greyidol** [comment]
You can customize the icon by placing an SVG in `app/javascript/images/logo.svg` and `app/javascript/images/app-icon.svg`, then running `RAILS_ENV=production bundle exec rails branding:generate`.

**Reply by Stoic42** [reply]
I have tried this but received Terrapin::CommandNotFoundError: Terrapin::CommandNotFoundError

**Reply by stevenleeg** [reply]
The `Terrapin::CommandNotFoundError` is likely due to your server not having `librsvg2-bin` installed (see the [branding.rake script](https://github.com/mastodon/mastodon/blob/main/lib/tasks/branding.rake#L31)). If you're on Ubuntu you can install this with `apt install librsvg2-bin`.
